### DISCUSSION TITLE
Number of units conditional on treatment

### DISCUSSION THREAD

**Discussion by gavin-kelly-1** [emitanaka/edibble#67] [Q&A]
Many thanks for the package @emitanaka - I can see this being a really useful tool for clarifying the process of designing experiments.  I'm wondering what is the colloquial way to handle the following situation.  Individual subjects receive a two-level treatment factor.  There's also a nested unit 'Cells' within the subjects - and subjects who get the first level of treatment will provide 2 nested Cells, but subjects who get the second level will provide 4 nested cells.

In the following code, I've attempted to nest a unit within a treatment so that I can use the 'RSV' value to determine the number of units, but attempting to serve the table gets a `Error in vapply(.x, .f, .mold, ..., USE.NAMES = FALSE) :   values must be length 1, but FUN(X[[1]]) result is length 0`. Nesting a unit within a treatment feels like it is abusing the grammar somewhat, but I am not familiar enough to find the correct way of phrasing this, so any advice would be very gratefully appreciated.

```
uv <- edibble::design() |>
  set_units(
    Subject=10
  ) |>
  set_trts(
    RSV=c("UV", "Infected")
  ) |>
  allot_trts(
    RSV~Subject
    ) |>
  set_units(
    Cells=nested_in(RSV, "UV" ~ 2, "Infected"~4)
  ) |>
  allot_units(Subject ~ Cells) |>
  assign_trts()

serve_table(uv)  
```

**Comment by emitanaka** [comment]
Thanks for the interesting challenge!

Currently, I don't allow nesting based on other variables. I can see this being useful, so I will try to implement this (perhaps something like in #68).

Regarding setting unit factors conditional on treatment -- originally, I reserved the helper functions `nested_in`, `crossed_by`, and `conditioned_on` for specifying within-role links only. In other words, I didn't want to allow users to set unit factors based on treatment factors and vice-versa. The reason for this is that any different procedure conditioned on treatments will then be confounded with the treatment allocation. This then can compromise the validity/interpretation of your experimental results. In your case, I'm wondering if it's worthwhile to mix up having 2 or 4 cells per subject (and don't condition on treatment)?

**Reply by gavin-kelly-1** [reply]
Thanks for the prompt answer - I agree with everything you've written.  I think a useful application of this package will be to transcribe confounded designs from a client's flawed reasoning, and then demonstrate their flaws through simulation, so I for one would appreciate the ability to produce invalid designs!

Regarding your final sentence, how would I 'mix up' the number of cells per subject.  I'm still getting used to the package so it's very likely I've missed an approach that is completely obvious and documented, so apologies for my ineptness. Other than writing out explicitly that subjects 1-5 have 2 cells and the rest have 4 (generate 15 formulae programmatically and then do.call them into the `nested_in` definition of cell units, which seems inelegant) I can't figure it out. 

(Just as background in case it's relevant in explaining my use-case: I suspect a problem is that the scientist behind this design is wanting to analyse two 'separate' datasets in one experimental design. UV treated subjects no longer generate one of the two cell types, so we can't assay it. So there are two natural questions: what is the difference in outcome (I've omitted the set_rcrd) between the cell-types in the non-UV-treated subjects, and what is the difference due to treatment in the cell-type that is common to treated and untreated.)

**Reply by emitanaka** [reply]
Okay, I'll allow for units conditional on treatment and built in some warning for this then (#69).
Regarding the specification, it is as you say "subjects 1-5 have 2 cells and the rest have 4" (you shouldn't have to use do.call):

```r
set_units(subject = 10, 
          cell = nested_in(subject, 
                    1:5 ~ 2,
                    6:10 ~ 4)) 
```

But this is clumsy and doesn't do randomisation! 

An alternative way, which I was thinking of for mixing-up, is to just encode it as data and not create the actual cell unit factor. This doesn't look desirable based on the context of your experiment though! (Thanks for providing the background!).

```r
set_units(subject = 10) %>%
  set_trts(ncell = c(2, 4)) %>%
  allot_table(ncell ~ subject)
```
